src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/NativeServerPushHookProcessor.java

```diff
@@ -0,0 +1,299 @@
+                    : new AbstractGitSCMSource.SCMRevisionImpl(head, change.getToHash());
+                result.put(head, revision);
+            }
+        }
+
+        private void addPullRequests(BitbucketSCMSource src, Map<SCMHead, SCMRevision> result) throws InterruptedException {
+            if (getType() != SCMEvent.Type.UPDATED) {
+                return; // adds/deletes won't be handled here
+            }
+
+            final BitbucketSCMSourceContext ctx = contextOf(src);
+            if (!ctx.wantPRs()) {
+                // doesn't want PRs, let the push event handle origin branches
+                return;
+            }
+
+            final String sourceOwnerName = src.getRepoOwner();
+            final String sourceRepoName = src.getRepository();
+            final BitbucketServerRepository eventRepo = refsChangedEvent.getRepository();
+            final SCMHeadOrigin headOrigin = src.originOf(eventRepo.getOwnerName(), eventRepo.getRepositoryName());
+            final Set<ChangeRequestCheckoutStrategy> strategies = headOrigin == SCMHeadOrigin.DEFAULT
+                ? ctx.originPRStrategies() : ctx.forkPRStrategies();
+
+            for (final NativeServerRefsChangedEvent.Change change : getPayload()) {
+                if (!"BRANCH".equals(change.getRef().getType())) {
+                    LOGGER.log(Level.INFO, "Received event for unknown ref type {0} of ref {1}",
+                        new Object[] { change.getRef().getType(), change.getRef().getDisplayId() });
+                    continue;
+                }
+
+                // iterate over all PRs in which this change is involved
+                for (final BitbucketServerPullRequest pullRequest : getPullRequests(src, change).values()) {
+                    final BitbucketServerRepository targetRepo = pullRequest.getDestination().getRepository();
+                    // check if the target of the PR is actually this source
+                    if (!sourceOwnerName.equalsIgnoreCase(targetRepo.getOwnerName())
+                        || !sourceRepoName.equalsIgnoreCase(targetRepo.getRepositoryName())) {
+                        continue;
+                    }
+
+                    for (final ChangeRequestCheckoutStrategy strategy : strategies) {
+                        if (strategy != ChangeRequestCheckoutStrategy.MERGE && !change.getRefId().equals(pullRequest.getSource().getRefId())) {
+                            continue; // Skip non-merge builds if the changed ref is not the source of the PR.
+                        }
+
+                        final String branchName = String.format("PR-%s%s", pullRequest.getId(),
+                            strategies.size() > 1 ? "-" + Ascii.toLowerCase(strategy.name()) : "");
+
+                        final PullRequestSCMHead head = new PullRequestSCMHead(branchName, sourceOwnerName,
+                            sourceRepoName, BitbucketRepositoryType.GIT, branchName, pullRequest, headOrigin, strategy);
```

Comment:
you should use `pullRequest.getSource().getBranch().getName()` instead of `branchName` otherwise it will not trigger the PR build (see how PullRequestSCMHead is created in `com.cloudbees.jenkins.plugins.bitbucket.BitbucketSCMSource#retrievePullRequests`)

see my comment on NativeServerPullRequestHookProcessor for the explanation